src/TransferList/TransferList.types.ts

```diff
@@ -1,14 +1,79 @@
-export type TransferListProps = {
+export type TransferListItem =
+  | string
+  | { primaryLabel: string; secondaryLabel?: string; id: string };
+
+export type TransferListProps<T> = {
+  /**
+   * The key to be used for the filter.
+   */
+  filterKey?: (item: T) => string;
+  /**
+   * Control the transfer of the items
+   */
+  handleTransfer?: (value: string[], intent: "toTarget" | "toSource") => void;
   /**
    * Array of Items.
    */
-  items?: string[];
+  items: TransferListItem[] | T[];
```

Comment:
Whats the use case for supporting a user defined item type? At the end of the day we only support having a primary and a secondary key. Just get the caller of the component to get the data in the right format rather than handling it here. Then we don't need to have `itemProps` to work out what the caller wants us to render.